Gems/EditorPythonBindings/Code/Source/PythonMarshalTuple.cpp

```diff
@@ -0,0 +1,309 @@
+        {
+            AZ_Error("python", false, "Tuple requires %zu elements but received %zu elements.", elements.size(), items.size());
+            return AZStd::nullopt;
+        }
+
+        // For each object found, create a copy of the value as the correct element in the C++ tuple.
+        // Also, save the pointers for each allocation that we do so that we can free them in the case of an error during conversion.
+        AZStd::vector<void*> reservedElements;
+        for (size_t itemIdx = 0; itemIdx < elements.size(); itemIdx++)
+        {
+            // Allocate space for each element.
+            void* reserved = tupleDataContainer->ReserveElement(tupleInstance.m_address, elements[itemIdx]);
+            AZ_Assert(reserved, "Could not allocate tuple's element %zu via ReserveElement()", itemIdx);
+
+            // Track the allocated space.
+            reservedElements.push_back(reserved);
+
+            // Attempt to convert the value. If it fails to convert, free everything we've allocated so far and return.
+            if (items[itemIdx] &&
+                !LoadPythonToTupleElement(
+                    items[itemIdx], traits, elements[itemIdx], tupleDataContainer, itemIdx, serializeContext, tupleInstance.m_address))
+            {
+                for (auto& reservedElement : reservedElements)
+                {
+                    tupleDataContainer->FreeReservedElement(tupleInstance.m_address, reservedElement, serializeContext);
+                }
+                return AZStd::nullopt;
+            }
+        }
+
+        outValue.m_value = tupleInstance.m_address;
+        outValue.m_typeId = tupleInstance.m_typeId;
+        outValue.m_traits = traits;
+
+        auto tupleInstanceDeleter = [behaviorClass, tupleInstance]()
+        {
+            behaviorClass->Destroy(tupleInstance);
+        };
+
+        return PythonMarshalTypeRequests::BehaviorValueResult{ true, tupleInstanceDeleter };
+    }
+
+    // Convert a C++ tuple into a Python list.
+    AZStd::optional<PythonMarshalTypeRequests::PythonValueResult> TypeConverterTuple::BehaviorValueParameterToPython(
+        AZ::BehaviorValueParameter& behaviorValue)
+    {
+        // the class data must have a container interface
+        AZ::SerializeContext::IDataContainer* containerInterface = m_classData->m_container;
+
+        if (!containerInterface)
+        {
+            AZ_Warning("python", false, "Container interface is missing from class %s.", m_classData->m_name);
+            return AZStd::nullopt;
+        }
+
+        if (!behaviorValue.ConvertTo(m_typeId))
+        {
+            AZ_Warning("python", false, "Cannot convert behavior value %s.", behaviorValue.m_name);
+            return AZStd::nullopt;
+        }
+
+        auto cleanUpList = AZStd::make_shared<AZStd::vector<PythonMarshalTypeRequests::DeallocateFunction>>();
+
+        // return tuple as list, if conversion failed for an element it will remain as 'none'
```

Comment:
Shouldn't this return a pybind tuple object: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#instantiating-compound-python-types-from-c ?

Reply:
happy to change it if you've got a preference. i considered it, but went with list to stay consistent with how pair<> marshaling is implemented. should be pretty easy to switch it though if you prefer.

Reply:
Ahhh, I see, thanks! You reserve the size on construction, but then you can put anything into it via array-style indexing without needing to use make_tuple(). Switching this to return a python tuple.